crates/gitbutler-repo/src/commands.rs

```diff
@@ -47,4 +49,26 @@ impl RepoCommands for Project {
         ctx.repository().remote(name, url)?;
         Ok(())
     }
+
+    fn read_file_from_workspace(&self, relative_path: &Path) -> Result<String> {
+        let ctx = CommandContext::open(self)?;
+        let base_path = ctx
+            .repository()
+            .path()
+            .parent()
+            .ok_or(anyhow::anyhow!("Could not find repository base path"))?;
+
+        let canonicalized_file_path = base_path.join(relative_path).canonicalize()?;
+
+        if canonicalized_file_path.as_path().starts_with(base_path) {
+            let tree = ctx.repository().head()?.peel_to_tree()?;
+            let entry = tree.get_path(relative_path)?;
+            let blob = ctx.repository().find_blob(entry.id())?;
+            let content = std::str::from_utf8(blob.content())?;
+
+            Ok(content.to_string())
+        } else {
+            anyhow::bail!("Invalid workspace file");
+        }
+    }
```

Comment:
Small nit, the `path()` function may return the repository path rather than the `.git` folder if the repository is bare. If we had a repository that was bare, that would mean that it could access files outside the repository.

```suggestion
    fn read_file_from_workspace(&self, relative_path: &Path) -> Result<String> {
        let ctx = CommandContext::open(self)?;
        let repository = ctx.repository();

        // If a repostiory is bare, repository.path will return the repository
        // directory rather than the `.git` repository.
        // Ref: https://docs.rs/git2/latest/git2/struct.Repository.html#method.path
        let base_path = if repository.is_bare() {
            repository.path()
        } else {
            repository
                .path()
                .parent()
                .ok_or(anyhow::anyhow!("Could not find repository base path"))?
        };

        let canonicalized_file_path = base_path.join(relative_path).canonicalize()?;

        if canonicalized_file_path.as_path().starts_with(base_path) {
            let tree = repository.head()?.peel_to_tree()?;
            let entry = tree.get_path(relative_path)?;
            let blob = repository.find_blob(entry.id())?;
            let content = std::str::from_utf8(blob.content())?;

            Ok(content.to_string())
        } else {
            anyhow::bail!("Invalid workspace file");
        }
    }
```


I wonder if we should name the function something like `read_commited_file` or something like that, to make it clear that its not going to look at uncommitted changes.

Reply:
https://docs.rs/git2/latest/git2/struct.Repository.html#method.path

Reply:
That's a great catch, @Caleb-T-Owens.
In general, there should be no assumption that `.git` is even in the worktree, independently of whether the repository is bare or not.
The only 'safe' way is to ask for the worktree-dir, and often the GitButler native types (like `Project` have a method for that.

Reply:
probably also worth noting that the function is trying to parse the file contents as utf8 - which of course wont work for files. I like the new blob reading implementation though, we can merge it and iterate from master

Reply:
Used `self.path` instead and double checked if the file is binary before tryign to read it as utf-8. 

I'll merge this then and like you said we can iterate from there :+1: